File: 2. Structural Design Patterns/Composite/Python/CompositeExample.py
# Composite Pattern Example in Python
# File system: File (leaf) and Directory (composite)

from abc import ABC, abstractmethod

class FileSystemItem(ABC):
    @abstractmethod
    def display(self, depth=0):
        pass
    
    @abstractmethod
    def get_size(self):
        pass

class File(FileSystemItem):
    def __init__(self, name, size):
        self.name = name
        self.size = size
    
    def display(self, depth=0):
        indent = "  " * depth
        print(f"{indent}File: {self.name} ({self.size}KB)")
    
    def get_size(self):
        return self.size

class Directory(FileSystemItem):
    def __init__(self, name):
        self.name = name
        self.items = []
    
    def add(self, item):
        self.items.append(item)
    
    def remove(self, item):
        self.items.remove(item)
    
    def display(self, depth=0):
        indent = "  " * depth
        print(f"{indent}Directory: {self.name}/")
        for item in self.items:
            item.display(depth + 1)
    
    def get_size(self):
        total = 0
        for item in self.items:
            total += item.get_size()
        return total

def main():
    root = Directory("root")
    docs = Directory("docs")
    docs.add(File("readme.txt", 5))
    docs.add(File("manual.pdf", 150))
    root.add(docs)
    root.add(File("config.ini", 2))
    
    root.display()
    print(f"Total size: {root.get_size()}KB")

if __name__ == "__main__":
    main()

File: 2. Structural Design Patterns/Composite/C#/CompositeExample.cs
// Composite Pattern Example in C#
// File system: File (leaf) and Directory (composite)
// Compile: csc CompositeExample.cs && CompositeExample.exe

using System;
using System.Collections.Generic;

namespace CompositeDemo
{
    abstract class FileSystemItem
    {
        public string Name { get; }
        protected FileSystemItem(string name) { Name = name; }
        public abstract void Display(int depth = 0);
        public abstract int GetSize();
    }

    class File : FileSystemItem
    {
        private readonly int _size;
        public File(string name, int size) : base(name) { _size = size; }
        public override void Display(int depth = 0) => Console.WriteLine(new string(' ', depth * 2) + $"File: {Name} ({_size}KB)");
        public override int GetSize() => _size;
    }

    class Directory : FileSystemItem
    {
        private readonly List<FileSystemItem> _items = new();
        public Directory(string name) : base(name) {}
        public void Add(FileSystemItem item) => _items.Add(item);
        public void Remove(FileSystemItem item) => _items.Remove(item);
        public override void Display(int depth = 0)
        {
            Console.WriteLine(new string(' ', depth * 2) + $"Directory: {Name}/");
            foreach (var item in _items) item.Display(depth + 1);
        }
        public override int GetSize() { int total = 0; foreach (var item in _items) total += item.GetSize(); return total; }
    }

    class Program
    {
        static void Main()
        {
            var root = new Directory("root");
            var docs = new Directory("docs");
            docs.Add(new File("readme.txt", 5));
            docs.Add(new File("manual.pdf", 150));
            root.Add(docs);
            root.Add(new File("config.ini", 2));
            root.Display();
            Console.WriteLine($"Total size: {root.GetSize()}KB");
        }
    }
}
